baseplate/lib/batched_queue.py

```diff
@@ -0,0 +1,45 @@
+from datetime import timedelta
+from queue import Queue
+from typing import Generic
+from typing import List
+from typing import Sequence
+from typing import TYPE_CHECKING
+from typing import TypeVar
+
+from baseplate.lib.timer import Timer
+
+T = TypeVar("T")
+
+if TYPE_CHECKING:
+    WorkQueue = Queue[Sequence[T]]  # pylint: disable=unsubscriptable-object
+else:
+    WorkQueue = Queue
+
+
+class BatchedQueue(Generic[T]):
+    """A queue which collects and placed items in batches."""
+
+    def __init__(self, work_queue: WorkQueue, batch_size: int, flush_interval: timedelta) -> None:
+        assert batch_size >= 1
+        assert work_queue is not None
+        self._work_queue = work_queue
+        self._batch: List[T] = []
+        self._batch_size = batch_size
+        self._flusher: Timer = Timer(self._flush, flush_interval)
+
+    def put(self, data: T) -> None:
+        self._batch.append(data)
+        if len(self._batch) >= self._batch_size:
+            self._flush()
+        else:
+            self._flusher.start()
+
+    def _flush(self) -> None:
+        batch = self.drain()
+        self._work_queue.put(batch)
+
+    def drain(self) -> List[T]:
+        self._flusher.stop()
+        batch = list(self._batch)
+        self._batch = []
+        return batch
```

Comment:
This feels pretty similar to https://github.com/reddit/baseplate.py/blob/develop/baseplate/sidecars/__init__.py#L55-L81

Rather than using threads, that solution relies on the pump timing out occasionally and putting an empty object into the batch to give it a chance to process timers. Can we do that here? I'd really like to avoid the threading stuff going on in the timer if we can.

Reply:
Yes, it's definitely similar in spirit. I've reworked it, so that the BatchedQueue does not flush automatically, but instead is called in the `on_iteration` handler.